stan/math/prim/meta/is_stan_closure.hpp

```diff
@@ -0,0 +1,85 @@
+#ifndef STAN_MATH_PRIM_META_IS_STAN_CLOSURE_HPP
+#define STAN_MATH_PRIM_META_IS_STAN_CLOSURE_HPP
+
+#include <stan/math/prim/meta/return_type.hpp>
+#include <stan/math/prim/meta/void_t.hpp>
+#include <stan/math/prim/meta/require_helpers.hpp>
+
+#include <type_traits>
+
+namespace stan {
+
+/**
+ * Checks if type is a closure object.
+ * @tparam The type to check
+ * @ingroup type_trait
+ */
+template <typename T, typename = void>
+struct is_stan_closure : std::false_type {};
+
+template <typename T>
+struct is_stan_closure<T, void_t<typename T::captured_scalar_t__>>
+    : std::true_type {};
```

Comment:
I was looking at the stanc3 pr, where is `captured_scalar_t__ defined in the closure?

Reply:
Closures are "duck-typed" so there's no single definition; every class has its own. But see for example [one_arg_closure](https://github.com/nhuurre/math/blob/9c2581759423fd61278a1f95467f3c45f7873a6d/stan/math/prim/functor/closure_adapter.hpp#L45-L46)